Step4 코드리뷰 요청드립니다.

lcjltj opened the pull request:
마지막 단계까지 우선 하긴했네요

리뷰어님께서 달아주신 피드백은 적용해보았고, 피드백에 대한 질문도 조금 남겨놨습니다.!

마지막 미션( 삭제 ) 부분도 꼼꼼한 피드백 부탁 드립니다.

lcjltj committed:
refactor: feedback 반영

lcjltj committed:
feat: 섹션(지하철역) 삭제 추가

jhhj424 commented on README.md:
```diff
@@ -60,5 +60,9 @@ npm run dev
       - [x] 새로운 역을 하행 종점 으로 등록
       - [x] 상행 역과 하행 역이 이미 노선에 모두 등록 되어 있을 경우 오류 발생
       - [x] 상행 역과 하행 역이 하나 라도 없을 경우 오류 발생
+  - [x] 구간 삭제 인수 테스트 작성
```
이번단계도 요구사항을 잘 작성해주셨습니다 ㅎㅎ 👍

jhhj424 commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -50,7 +49,7 @@ public LineResponse fetchLine(final Long id) {
     public void updateLine(final Long id, final LineUpdateRequest lineUpdateRequest) {
         final Line line = findById(id);
-        line.updateNameAndColor(lineUpdateRequest);
+        line.updateNameAndColor(lineUpdateRequest.getName(), lineUpdateRequest.getColor());
```
👍

jhhj424 commented on src/main/java/nextstep/subway/line/application/SectionService.java:
```diff
@@ -28,4 +28,12 @@ public LineResponse addSection(final Long id, final SectionAddRequest sectionAdd
     }
+
+    @Transactional
+    public void deleteSection(final Long id, final Long stationId) {
```
```suggestion
    public void deleteSection(final Long lineId, final Long stationId) {
```

아마 당연히 알고계실테지만 체크 못하신 것 같아서 말씀드립니다 :)
조금 더 명확하게 하면 좋을 것 같아서요 👍

jhhj424 commented on src/main/java/nextstep/subway/line/domain/Line.java:
```diff
@@ -66,11 +67,8 @@ public String getColor() {
-                .map(StationResponse::of)
-                .collect(Collectors.toSet());
+    public Set<Station> getAllStations() {
+        return sections.getAllStation();
```
질문해주신 내용에 포함 된 부분인 것 같은데 :)
외부에서 해당 컬렉션을 조작할 수 없도록 불변 컬렉션으로 `Return` 하는 방법도 좋을 것 같아요 ㅎㅎ

참고 : https://tecoble.techcourse.co.kr/post/2021-04-26-defensive-copy-vs-unmodifiable/

---

보통은 일급컬렉션이 가지고 있는 element 에 대해서 해당 방식을 사용하곤 하니, Sections 클래스의 getAllStation()메서드에서 불변 컬렉션을 return 하는게 어떨까요?

jhhj424 replied:
### 질문 관련해서 여기서 코멘트 남길게요 !!

---

```
리뷰어님! 의말대로 DTO(request, repsonse)에 의존이 생기는 부분을 제거하였습니다.
```

일단 엔티티에서 DTO 에 의존하는 부분은 제거 잘 해주셨습니다. 이를 유도한 이유는 앞서 얘기한 맥락과 동일한데 결국, 엔티티는 외부의 요구사항(특정 view 스펙 등)에 의해서 변경되지 않는게 중요하다고 생각해서입니다 :)

```
sections 관련 하여 , allStations 를 리턴 해주는 메소드 사용을. 도메인에 그대로 List을 리턴해주려고 하고있습니다.
getSections으로 sections를 외부에 노출 시키게 되면 해당 sections을 변경 할 수 있는 범위가 너무 커질거 같아 그렇게 하려고 작성을 하였는데
```

`allStations()` 관련해서는 위에 댓글을 참고해주시면 좋을 것 같습니다 !!

```
 또 하나의 질문은 컨트롤러에서 서비스에 dto를 넘겨주는 것입니다.
서비스를 dto를 넘기게되면 해당 서비스 메소드는 해당 해당 dto로만 받아야 해 재사용이 많이 떨어진다고 생각합니다 ㅠㅠ..
리뷰어님 생각은 어떠신가요?
```

컨트롤러에서 서비스에 dto(request) 를 넘겨주시는걸 말씀해주신걸까요?
일단 제가 이해한걸 바탕으로 지극히 개인적인 의견을 말씀드리겠습니다 :)

제가 생각하는 `서비스`에서는 각자의 도메인 로직(핵심 비즈니스 로직)을 가지는 `도메인을 조합`해서 특정 기능을 구현하는 레이어로 생각하고있습니다.
일단 서비스에서의 메서드는 구현에 필요한 최소한의 정보들만 외부에서 받도록 하는데 이것을 dto(request) 로 현재 사용하는것이고, 같은 구현에서는 거의 대부분 같은 외부 데이터를 요구한다고 생각하고 있습니다.
그래서 말씀해주신대로 재사용성 측면에서 개선하려면 현재 구조처럼 `api endpoint` 에서 받은 `request` 를 바로 서비스에 넘겨서 사용하는 구조가 아니고, 한번 더 `service layer` 용 `dto` 로 변환해서 사용하면 어떨까 하는데, 어떻게 생각하실까요? :)

lcjltj replied:
넵!  위에 내용은 리뷰어님 말씀도 그렇고 스터디 하는 친구들과 얘기한 주제이긴한데 서비스 레이어용 dto로 변환해서 하는게 더 효율적일수 있다고 생각해요.  api 요청으로 온 데이터말고 외부 에서 받아온데이터를 서비스에서 필요로 하는 서비스 도 있어서 그렇게 하는 경우도 많다고 하더라구요.!!

jhhj424 commented on src/test/java/nextstep/subway/line/domain/SectionsTest.java:
```diff
@@ -82,6 +82,35 @@ void setUp() {
+    @Test
+    void 섹션_종착역_삭제() {
+        // given
+        Section newSection = Section.of(StationTest.강남역, StationTest.이수역, 4L);
```
`Section newSection = Section.of(StationTest.강남역, StationTest.이수역, 4L);`

이러한 반복되는 동작들을 `Fixture` `클래스` 또는 `메서드` 단위로 분리해서 사용하면 `가독성`, `재사용성`을 더 확보할 수 있을 것 같은데 어떻게 생각하시나요? :)

lcjltj replied:
반복되는 부분이 많아.. 보기 안좋긴하네요 개선하게씁니다!

jhhj424 commented on src/test/java/nextstep/subway/line/SectionAcceptanceTest.java:
```diff
@@ -45,6 +44,7 @@ void setUp() {
         }
         databaseCleanup.execute();
         지하철역_추가();
+        line = lineRepository.save(Line.of("2호선", "bg-blue-600", 10L, StationTest.강남역, StationTest.사당역));
```
인수 테스트 관련해서 `인수 테스트의 의도 드러내기` 강의 내용이 있었는데, 해당 부분에서 각 스텝들을 한글 메서드를 사용하도록 해서 의도를 더 드러내도록 했었는데, 이러한 부분도 적용될 수 있을 것 같아요 :)

아무래도 인수테스가 시나리오를 검증하는거다보니 다른 사람들이 봤을때도 정말 쉽게 시나리오가 읽히고 의도가 드러나면 좋을 것 같다는 생각입니다 !

---

다음 미션중에 다시 인수테스트를 사용하는 미션들이 있는데 이번 미션 강의 예시처럼, 이런식으로 명확하게 나눠서 작성해보시는 연습을 하는것도 좋을 것 같아요 ~
<img width="970" alt="image" src="https://user-images.githubusercontent.com/46177529/173608139-f0dede79-ab23-4a15-9c51-6f267ed62b87.png">

lcjltj replied:
아무레도 한글로 변수명을 하는게 익숙하지가 않아서 그런거같아요.!
최대한 테스트케이스 작성할땐 의미전달이 잘 되도록 한글로 하도록하겠습니다.

jhhj424 commented on src/test/java/nextstep/subway/line/SectionAcceptanceTest.java:
```diff
@@ -45,6 +44,7 @@ void setUp() {
         }
         databaseCleanup.execute();
```
이전 단계에서 한번 말씀드렸을 것 같은데요 :)

`databaseCleanup.execute();` 와 같은 인수테스트에서 항상 사용하는 동작들을 모아놓은 클래스로 분리해서 사용할 수 있을 것 같아요~
그렇게 되면 어떤 인수테스트를 작성하든 해당 클래스를 상속받아서 기본 인수테스트 `setUp` 을 공통으로 처리할 수 있다는 장점이 생길 것 같습니다 👍

lcjltj replied:
인수테스트  클래스 생성 시 공통으로 사용 되는 부분들을 공통 클래스를 만들어서 분리 했습니다.!!

jhhj424 reviewed (changes requested):
안녕하세요 찬준님~ 🙇🏻‍♂️
인수 테스트 주도 개발 미션 4단계 잘 작성해주셨습니다 💯 
질문에 대한 답변과 함께 마지막으로 조금 더 진행했으면 좋을 것 같은 부분 코멘트 남겼으니 확인하시고, 반영하실 수 있는 부분은 추가로 반영해보는게 어떨까요 ? 😊
앞으로의 미션에도 도움이 되실거라 생각하고 `Request changes` 한번만 드리겠습니다 🙏
미션 진행하시느라 정말 고생 많으셨습니다 !! 마지막까지 화이팅입니다 👊

lcjltj committed:
refactor: feedback 반영

- AcceptanceTest 생성 및 공통 기능 분리 / 각 인수 테스트 상속
- 파라메터 변수명 변경 id -> lineId
- 섹션 추가 메소드 생성 (중복 제거)

lcjltj committed:
refactor: 콜렉션 관련 방어적 복사로 변경

lcjltj committed:
fix: 테스트 케이스 오류 수정

jhhj424 reviewed (approved):
안녕하세요 찬준님~ 🙇🏻‍♂️
인수 테스트 주도 개발 미션 마지막 피드백까지 너무 잘 반영해주셨습니다 💯 
미션중에 던져주신 질문에 대해 저도 고민해볼 수 있는 시간이 있어서 좋았습니다 !
이번 미션 진행하시느라 정말 고생 많으셨습니다 !! 
언제든 궁금하신 점 있으시면 DM 주세요 ㅎㅎ 마지막까지 화이팅입니다!! 👊
